src/lib/operators/table_scan/column_vs_column_table_scan_impl.cpp

```diff
@@ -52,17 +52,35 @@ std::shared_ptr<PosList> ColumnVsColumnTableScanImpl::scan_chunk(ChunkID chunk_i
   /**
    * FAST PATH
    * ...in which the SegmentType does not get erased in Release builds. DataTypes and SegmentTypes have to be the same.
+   * In the case of ReferenceSegments, the used ReferenceSegmentIterator has to be the same (see below).
    */
   if (left_segment->data_type() == right_segment->data_type()) {
     resolve_data_and_segment_type(*left_segment, [&](auto data_type_t, auto& left_typed_segment) {
       using ColumnDataType = typename decltype(data_type_t)::type;
       using SegmentType = std::decay_t<decltype(left_typed_segment)>;
 
       if (const auto right_typed_segment = std::dynamic_pointer_cast<SegmentType>(right_segment)) {
-        // Same segment types - do not erase types in Release builds
-        result = _typed_scan_chunk<EraseTypes::OnlyInDebugBuild>(
-            chunk_id, create_iterable_from_segment<ColumnDataType>(left_typed_segment),
-            create_iterable_from_segment<ColumnDataType>(*right_typed_segment));
+        if constexpr (std::is_same_v<SegmentType, ReferenceSegment>) {
+          // For reference segments check if the iterables end up using the same type of iterator.
+          auto left_iterable = ReferenceSegmentIterable<ColumnDataType, false>{left_typed_segment};
+          auto right_iterable = ReferenceSegmentIterable<ColumnDataType, false>{*right_typed_segment};
+
+          left_iterable.with_iterators([&](auto left_it, const auto left_end) {
+            right_iterable.with_iterators([&](auto right_it, const auto right_end) {
+              if constexpr (std::is_same_v<std::decay_t<decltype(left_it)>, std::decay_t<decltype(right_it)>>) {
```

Comment:
Just a comment: The lengths that we need to go to to save compile time are getting out of hand. We should drop the type inter-operability, it's just not achievable in a static-polymorphic world...

I mean, look at this, how many people in the radius of 50km around HPI can decipher what we're doing here? :(